designs/0035-cmdstanrs.md

```diff
@@ -0,0 +1,348 @@
+- Feature Name: cmdstanrs
+- Start Date: 2023-11-22
+- RFC PR:
+- Stan Issue:
+
+# Summary
+[summary]: #summary
+
+This is a proposal for a Rust interface for CmdStan through compiled
+executables, that is, no direct interface with C++.
+
+The goal is to provide an interface which enables users to:
+- compile Stan programs (with arbitrary options)
+- build and compose arguments/options (to be passed to C++
+  executables) in idiomatic Rust
+- call C++ executables, then memoize input and collect output (thereby
+  making these available for programmatic use)
+- call `diagnose` and `stansummary` tools and collect output
+
+The objective is to keep the interface as simple as possible.
+
+# Motivation
+[motivation]: #motivation
+
+Suppose that you write Rust code. Suppose that you use `Stan` for
+probabilistic programming. You have three choices for creating an
+application which utilizes both: shell scripts, introduce a scripting
+language (e.g. Python) dependency, or a Rust interface for `CmdStan`.
+
+Orchestration using the shell suffers from portability issues and
+often leads unnecessary fragmentation of control flow. Introducing a
+scripting language may confer portability, but control flow and error
+handling are now divided between two languages; furthermore, this
+necessitates code be written to serialize/deserialize intermediates.
+
+A Rust interface, in similar spirit to the CmdStan interfaces from
+other languages, would provide the necessary abstraction to eliminate
+the aforementioned problems.
+
+# Functional Specification
+
+Given a Stan program, a user of the library will compile the model (if
+desired), call the executable with arguments (translated from
+strongly-typed argument tree), and obtain a self-contained context
+which encapsulates the pertinent information from the call.
+
+### Assumptions
+
+We assume (at our peril) that Rust programmers that will be able to
+figure out how to satisfy the following requirement:
+- a working CmdStan installation exists at some user-accessible path
+
+### Processes, IO
+
+The proposal is to use the Rust `std` library, in particular the
+[process](https://doc.rust-lang.org/std/process/index.html),
+[path](https://doc.rust-lang.org/std/path/index.html),
+[fs](https://doc.rust-lang.org/std/fs/index.html), and
+[ffi](https://doc.rust-lang.org/std/ffi/index.html) modules, to
+orchestrate processes, interact with file system and handle
+cross-platform concerns. This will yield a library which is portable,
+provided that it is (cross-)compiled for the intended target.
+
+## Control: compilation and calling the resultant executable
+
+A `CmdStanModel` type will serve as an abstraction for a Stan program,
+which may need to be compiled. Rather than compile on construction, a
+compilation method must be explicitly called in user code (assuming
+that a satisfactory executable does not exist yet).
```

Comment:
This model is something CmdStanPy [is moving away from](https://github.com/stan-dev/cmdstanpy/issues/695), because a CmdStanModel is pretty useless without an executable. 

In a strongly typed context, one could imagine that this type could instead be called something like `StanProgram`, and the result of calling `compile` is a `CmdStanModel` which is always associated with a executable on disk - this is sometimes called the "[typestate](http://cliffle.com/blog/rust-typestate/)" pattern

Reply:
Excellent idea. As is often the case, putting a name to a common idiom (the typestate pattern) is illuminating.